### ISSUE TITLE
Integration with `python_mattermost_driver`

### ISSUE THREAD

**Issue by jneeven** [attzonko/mmpy_bot#148]
Hi, first of all, thanks a lot for creating this python package! The mattermost bot ecosystem is appalling, and being able to do things in python just makes everything so much easier :sweat_smile: Over the last few days I've been looking at the code a bit and fiddling with it here and there, and while there seems to be a lot of functionality here already, much of the API is also missing (which makes sense, because why implement it if you don't use it).

I was just wondering; have you thought about delegating all the API functionality to https://github.com/Vaelor/python-mattermost-driver/? It's actively maintained and has almost the entire range of mattermost API endpoints implemented as python functions. I think this would be a great opportunity to both add functionality and clean the code up a lot! The only issue might be that it's python 3 only, and this package also supports python 2. However, now that [python 2 itself is no longer maintained](https://www.python.org/dev/peps/pep-0373/#update), I'm not sure if that really matters?

Curious to hear your thoughts on this matter, I would definitely be willing to help refactor.

**Comment by attzonko** [comment]
Thanks for the kind words. Honestly, I was not the original author of this code, I inherited it when the original author stopped maintaining it. We did some upgrades when the API version went from 3 to 4, but it has been on cruise control for a while since I don't have a lot of spare time for improvements. 

I have seen the driver, and also thought it would be a good idea to use it, but as I mentioned I personally have had a hard time finding the bandwidth to do much development. I agree that we have had more than enough time maintaining Python 2 backward compatibility. 

I would not be opposed if the community decided to put in a PR deprecating Python 2 support and refactoring the code to utilize the driver.

**Comment by jneeven** [comment]
Awesome, thanks for the response! I already started refactoring locally, but since it touches almost all parts of the code (and hence either breaks existing code or makes it obsolete) I found it easier to start somewhat from scratch and just copy in the (python 3 only) elements I needed.

I have a hard-fork, heavily simplified refactor going in github.com/jneeven/snaketalk, so I will keep building on that until the basic functionality is there again, and then perhaps we can have a look at integrating the two repositories :+1:

**Comment by ttuffin** [comment]
Good initiative @jneeven :+1: I started looking into implementing asyncio with mmpy_bot as requested in #124, but have not had much time to focus on it. I like your idea of integrating with [python_mattermost_driver](https://github.com/Vaelor/python-mattermost-driver/) since it is actively maintained and would make sense to use their existing library instead of rewriting and maintaining more code. It looks like they use asyncio for the websockets connection but not for anything else, and I am curious what benefits asyncio and aiohttp could bring to the bot as was mentioned in #124. 

I'll keep a close eye on your fork and submit some PR's/ideas should any pop-up, and as you say, hopefully we could reach a point where we can integrate the two projects.

**Comment by jneeven** [comment]
@ttuffin my fork is now pretty much fully working; it just needs tests to ensure it keeps working in the future (I'll wait with releasing any kind of python package until those are in place). It's still missing some minor functionality (such as scheduling, a help function and triggering webhooks, but all code from the example plugins is working and it has full asyncio support, should you need it. I did build it from the ground up with this repo as an inspiration, so the overall structure of e.g. how plugins work is a bit different.

@attzonko let me know at what point you think we might want to start copying changes to this repo! In the meantime I will continue developing over there.

**Comment by attzonko** [comment]
@jneeven I took a look at your fork and like the direction you took with the refactor. I agree that there is some functionality missing, but I would be fine with you migrating the refactor to this repo. One feature I did not see (unless I just rushed through the code), which I think is needed is the ability to authenticate the bot with a token. I think that with newer versions of Mattermost that is a must. I do think we should prioritize getting the tests working as well 😺 Let me know if you need anything from me on this end.

**Comment by jneeven** [comment]
> One feature I did not see (unless I just rushed through the code), which I think is needed is the ability to authenticate the bot with a token.

Token authentication is currently the only supported way (see [here](https://github.com/jneeven/snaketalk/blob/main/snaketalk/bot.py#L34)), since I deemed it unlikely that anyone would ever want to use this with an actual user account, but support for that can be implemented in a matter of seconds if necessary. I agree that the tests are top priority atm (I put them at the top of the roadmap as well), so if anyone has time to write any kind of test at all, any help would be greatly appreciated!

**Comment by attzonko** [comment]
@jneeven Ok Travis CI is back to a working state thanks to @rwky so I think we can safely try and bring your refactor fork over here if you are still interested.

**Comment by jneeven** [comment]
> @jneeven Ok Travis CI is back to a working state thanks to @rwky so I think we can safely try and bring your refactor fork over here if you are still interested.

Definitely! The repo is shaping up quite nicely, have a look at https://github.com/jneeven/snaketalk/pulls. There are still two todo items (webhook server and click functions) but even once the current PRs are merged I'd be ready to get things moving here. I've also re-done the entire testing setup from scratch; my repo uses a mattermost server running inside of a docker container to do automated integration tests, and it seems to work quite nicely. Maybe you could have a look at the current state of the repo (let me know if those PRs need to be merged first, I'm basically waiting if @ttuffin wants to have a look, but I'd be fine with merging them without review) and let me know if there are any major features you think need to be added before we can get things going over here?

**Comment by attzonko** [comment]
@jneeven I don't think you need to wait for those two PRs. In fact, the change that @rwky did runs the current tests using docker in TravisCI which I hope should make the merge even easier. Let me know if you need anything from me.

**Comment by jneeven** [comment]
> @jneeven I don't think you need to wait for those two PRs. In fact, the change that @rwky did runs the current tests using docker in TravisCI which I hope should make the merge even easier. Let me know if you need anything from me.

How do you want to go about this? Should I just make a PR replacing *all* files that we can then make other PRs to (e.g. to update the docs)? Because I pretty much started from scratch, I think it'll be difficult to simply change the existing files (including the tests)

**Comment by ttuffin** [comment]
Based on @attzonko's [comment](https://github.com/attzonko/mmpy_bot/issues/124#issuecomment-780876768), perhaps we need a new branch for mmpy_bot 2.0 and then merge snaketalk into there. Once things are stable and 2.0 is ready for release then it could be merged into the master branch. If the 1.x stream will still be supported then I guess we would need to move that to it's own branch before merging 2.0 into master?
